src/main/java/org/jenkinsci/plugins/workflow/log/DelayBufferedOutputStream.java

```diff
@@ -120,6 +116,45 @@ void flushAndReschedule() {
 
     }
 
+    /**
+     * Flushes streams prior to garbage collection.
+     * ({@link BufferedOutputStream} does not do this automatically.)
+     * TODO Java 9+ could use java.util.Cleaner
+     */
+    private static final class FlushRef extends PhantomReference<DelayBufferedOutputStream> {
+
+        private static final ReferenceQueue<DelayBufferedOutputStream> rq = new ReferenceQueue<>();
+
+        static {
+            Timer.get().scheduleWithFixedDelay(() -> {
+                while (true) {
+                    FlushRef ref = (FlushRef) rq.poll();
+                    if (ref == null) {
+                        break;
+                    }
+                    LOGGER.log(Level.FINE, "cleaning up phantom {0}", ref.out);
+                    try {
+                        ref.out.flush();
```

Comment:
Why flush the underlying stream instead of the `DelayBufferedOutputStream` itself? Couldn't there be data left in the buffer that wouldn't get flushed?

Reply:
It is not possible to get to the `DelayBufferedOutputStream` from a finalizer. But yes, this will mean `flushBuffer()` is not called. I think the only way to fix that would be to turn `DelayBufferedOutputStream` into a `FilterOutputStream` that delegates to a plain `BufferedOutputStream`.